### PR TITLE
Provider Databricks add jobs create operator.

### LINKED ISSUE #29733: Databricks create/reset then run-now
### Description

Allow an Airflow DAG to define a Databricks job with the `api/2.1/jobs/create` (or `api/2.1/jobs/reset`) endpoint then run that same job with the `api/2.1/jobs/run-now` endpoint. This would give similar capabilities as the DatabricksSubmitRun operator, but the `api/2.1/jobs/create` endpoint supports additional parameters that the `api/2.1/jobs/runs/submit` doesn't (e.g. `job_clusters`, `email_notifications`, etc.).

### Use case/motivation

Create and run a Databricks job all in the Airflow DAG. Currently, DatabricksSubmitRun operator uses the `api/2.1/jobs/runs/submit` endpoint which doesn't support all features and creates runs that aren't tied to a job in the Databricks UI. Also, DatabricksRunNow operator requires you to define the job either directly in the Databricks UI or through a separate CI/CD pipeline causing the headache of having to change code in multiple places.

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by kyle-winkelman**
Add the DatabricksJobsCreateOperator for use cases where the DatabricksSubmitRunOperator is insufficient.
closes: #29733

### CODE REVIEW

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
cc: @alexott ?

**Comment by alexott** [comment]
@kyle-winkelman what are the use cases for that operator?

**Comment by kyle-winkelman** [comment]
My specific use case is that I want to run multiple tasks in the same job and on a single `job_clusters`. This is not supported by the DatabricksSubmitRunOperator because it doesn't support the parameter `job_clusters`, so the only way to support it is to create a new_cluster per task I want to run.

The other approach is to use the DatabricksRunNowOperator which has the limitation that you have to define your Databricks Job somewhere else and in some different manner (i.e. manually in Databricks UI, custom CI/CD pipeline, etc.). My team doesn't like having the definition of the job be separated from the use of it from Airflow. In my opinion the DatabricksRunNowOperator with just a single `job_id` lacks information about what is actually happening.

So to sum up, it is useful to be paired with the DatabricksRunNowOperator to define a job and run it in the same DAG.

File: airflow/providers/databricks/operators/databricks.py [hunk 1]
```diff
@@ -151,6 +151,145 @@ def get_link(
...
+        tags: dict[str, str] | None = None,
+        tasks: list[object] | None = None,
+        job_clusters: list[object] | None = None,
+        email_notifications: object | None = None,
+        webhook_notifications: object | None = None,
+        timeout_seconds: int | None = None,
+        schedule: dict[str, str] | None = None,
+        max_concurrent_runs: int | None = None,
+        git_source: dict[str, str] | None = None,
+        access_control_list: list[dict[str, str]] | None = None,
```
Comment (alexott) [inline airflow/providers/databricks/operators/databricks.py:226]: For new operators I would really like to use real data structures (for example, data classes), not the simple objects as it doesn't provide users with auto-completion, etc. - it's easy to make mistake in the untyped JSON definition

File: airflow/providers/databricks/operators/databricks.py [hunk 2]
```diff
@@ -151,6 +151,145 @@ def get_link(
...
+        )
+
+    def execute(self, context: Context) -> int:
+        self.job_id = self.xcom_pull(
+            context,
+            task_ids=self.task_id,
+            include_prior_dates=True,
+        )
+        if self.job_id:
+            self._hook.reset(self.job_id, self.json)
```
Comment (alexott) [inline airflow/providers/databricks/operators/databricks.py:287]: Instead of I would recommend to get the current job definition and compare with the new definition, and reset only if this definition changes (that should happen relatively rare). 

Also, you need to handle following edge cases:

* Job is deleted via UI - the reset will fail because job ID doesn't exist
* We can lose XCom information, so we'll create a duplicate for the job.

**Review by alexott — COMMENTED** [review]
I'll discuss this further with the Databricks Workflows team

File: airflow/providers/databricks/operators/databricks.py (hunk 1 above)
Comment (kyle-winkelman) [reply airflow/providers/databricks/operators/databricks.py:226]: Do you have an example of a different operator that follows this pattern? It would be helpful to see an example.

Any thoughts on a tool to generate such data structures from the [Databricks OpenAPI Spec](https://docs.databricks.com/_extras/api-refs/jobs-2.1-aws.yaml)? There are a lot of data structures that would need to be created and I don't want to do so manually.

My initial thought was to rely on [Databricks Python API](https://docs.databricks.com/dev-tools/python-api.html), but it doesn't have these data structures or validations either.

Comment (alexott) [reply airflow/providers/databricks/operators/databricks.py:226]: I don’t have existing example, it was just thoughts for the new operators. New Python API will be available soon, that will provide access to the latest APIs. I need to ask dev team when it’s coming

Comment (kyle-winkelman) [reply airflow/providers/databricks/operators/databricks.py:226]: Do you think this new operator should wait for the new Python API that you expect to be available soon?

Comment (alexott) [reply airflow/providers/databricks/operators/databricks.py:226]: I’m waiting for answer from product management - maybe we won’t need this operator…

Comment (kyle-winkelman) [reply airflow/providers/databricks/operators/databricks.py:226]: Have you heard anything from the Databricks product management @alexott?

File: airflow/providers/databricks/operators/databricks.py (hunk 2 above)
Comment (alexott) [reply airflow/providers/databricks/operators/databricks.py:287]: not yet

File: airflow/providers/databricks/operators/databricks.py (hunk 1 above)
Comment (alexott) [reply airflow/providers/databricks/operators/databricks.py:226]: no decision yet.

Comment (eladkal) [reply airflow/providers/databricks/operators/databricks.py:226]: @alexott lets please get to a decision in the upcoming days (before next wave of providers)... if not i will accept the PR as is. 
Should in the future Databricks decide against these operators and present arguments why we shouldn't have them we can always remove with a major release.

**Review by potiuk — CHANGES_REQUESTED** [review]
@alexott convinced me that it needs changes.

**Comment by potiuk** [comment]
> @alexott convinced me that it needs changes.

@alexott - are there any news from Databricks on that one? We are considering merging it regardless, but it seems you wanted to have some input from the team first?

**Comment by alexott** [comment]
@potiuk Unfortunately no news. The decision takes longer than planned. If we can deprecate it in the later versions, let merge it.

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
> @potiuk Unfortunately no news. The decision takes longer than planned. If we can deprecate it in the later versions, let merge it.

OK. Cool. I am a big fan of having tactical solutions that solve part of the probem or a problem for smaller group of people  and releasing them early, as long as they are not preventing strategic long term solutions that needs a longer debate and far more work to be implemented later. This one looks like one of those.

I re-reviewed it and it looks cool. @kyle-winkelman -> can you please rebase and fix the static check problem, then I am happy  to merge it (@eladkal  - WDYT?).

**Comment by lennartkats-db** [comment]
@potiuk @kyle-winkelman This seems like a useful addition from the Databricks perspective. Thanks for contributing, Kyle! 🙏

I do think it would be good to look at the XCOM issue that Alex called out. Wouldn't it be better to just use the API to find an existing job for the cases that Alex called out where the current approach doesn't work? To avoid creating duplicate jobs or having the operator fail? The `DatabricksRunNowOperator` already applies [logic like this](https://github.com/apache/airflow/blob/91568b26a448bf15c91dcb4497690d54842678fc/airflow/providers/databricks/hooks/databricks.py#L211), as it allows users to run a given job based on the name of that job (rather than the job id).

**Comment by stikkireddy** [comment]
hey @kyle-winkelman do you have bandwidth to move forward with this PR?

**Comment by stikkireddy** [comment]
@potiuk is it okay if someone else forks Kyle's repo and fix the static check to not lose @kyle-winkelman's credit/commits and also address the previous comments made by @alexott? I am also okay to wait a bit for Kyle to respond.

**Comment by potiuk** [comment]
> @potiuk is it okay if someone else forks Kyle's repo and fix the static check to not lose @kyle-winkelman's credit/commits and also address the previous comments made by @alexott? I am also okay to wait a bit for Kyle to respond.

Fine for me.
